### ISSUE TITLE
bcftools +fixref -- -m top is buggy

### ISSUE THREAD

**Issue by freeseek** [samtools/bcftools#825]
There is a bug in `plugins/fixref.c` in the determination of the reference forward strand as either <a href="https://www.illumina.com/documents/products/technotes/technote_topbot.pdf">TOP or BOT</a>:
```
for (i=1; i<=win; i++)
{
    int ra = nt2int(ref[mid-i]);
    int rb = nt2int(ref[mid+i]);
    if ( ra<0 || rb<0 || ra==rb ) continue;     // skip N's and non-infomative pairs: A/A, C/C, G/G, T/T
    int pair = 1 << ra | 1 << rb;
    if ( pair==0x9 || pair==0x6 ) continue;     // skip ambiguous pairs: A/T or C/G
    strand = ra & 0x9 ? 1 : -1;
    break;
}
```
should be changed to
```
for (i=1; i<=win; i++)
{
    int ra = nt2int(ref[mid-i]);
    int rb = nt2int(ref[mid+i]);
    if ( ra<0 || rb<0 || ra==rb ) continue;     // skip N's and non-infomative pairs: A/A, C/C, G/G, T/T
    int pair = 1 << ra | 1 << rb;
    if ( pair==0x9 || pair==0x6 ) continue;     // skip ambiguous pairs: A/T or C/G
    strand = 1 << ra & 0x9 ? 1 : -1;
    break;
}
```

Here below is code to reproduce the issue:
```
(echo "##fileformat=VCFv4.2"
echo "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">"
echo "##contig=<ID=1>"
echo "##contig=<ID=2>"
echo "##contig=<ID=3>"
echo "##contig=<ID=4>"
echo "##contig=<ID=5>"
echo "##contig=<ID=6>"
echo "##contig=<ID=7>"
echo "##contig=<ID=8>"
echo -e "#CHROM\tPOS\tID\tREF\tALT\tQUAL\tFILTER\tINFO\tFORMAT\tSAMPLE"
echo -e "1\t11\trs1535632\tA\tT\t.\t.\t.\tGT\t0/0"
echo -e "2\t11\trs363334\tC\tG\t.\t.\t.\tGT\t0/0"
echo -e "3\t11\trs7101540\tA\tT\t.\t.\t.\tGT\t0/0"
echo -e "4\t11\trs7113791\tA\tT\t.\t.\t.\tGT\t0/0"
echo -e "5\t11\trs778833\tC\tG\t.\t.\t.\tGT\t0/0"
echo -e "6\t11\trs4933195\tA\tT\t.\t.\t.\tGT\t0/0"
echo -e "7\t11\trs903997\tC\tG\t.\t.\t.\tGT\t0/0"
echo -e "8\t11\trs1942968\tC\tG\t.\t.\t.\tGT\t0/0") > file.vcf

(echo ">1"
echo "ACGGGGACAGATATGTTAACT"
echo ">2"
echo "ATGAGTGAATCAAGCACTATT"
echo ">3"
echo "AAATTCAGATACAGAATCTTT"
echo ">4"
echo "GATGGACAGGATGACCTCTAG"
echo ">5"
echo "GGTTAAATGTCAAGGTGAGCT"
echo ">6"
echo "ATGCTAATAAAACATTAAAGT"
echo ">7"
echo "ATGAGAAAGTCTGAGAGTGCA"
echo ">8"
echo "CTACATGACTCTTTATGTTAC") > ref.fa

samtools faidx ref.fa

bcftools +fixref file.vcf -- -f ref.fa -m top
```

The buggy output is this:
```
1	11	rs1535632	A	T	.	.	.	GT	1/1
2	11	rs363334	C	G	.	.	.	GT	1/1
3	11	rs7101540	A	T	.	.	.	GT	0/0
4	11	rs7113791	A	T	.	.	.	GT	1/1
5	11	rs778833	C	G	.	.	.	GT	1/1
6	11	rs4933195	A	T	.	.	.	GT	1/1
7	11	rs903997	C	G	.	.	.	GT	1/1
8	11	rs1942968	C	G	.	.	.	GT	0/0
```

The output with the fix is this:
```
1	11	rs1535632	A	T	.	.	.	GT	1/1
2	11	rs363334	C	G	.	.	.	GT	0/0
3	11	rs7101540	A	T	.	.	.	GT	0/0
4	11	rs7113791	A	T	.	.	.	GT	1/1
5	11	rs778833	C	G	.	.	.	GT	1/1
6	11	rs4933195	A	T	.	.	.	GT	0/0
7	11	rs903997	C	G	.	.	.	GT	0/0
8	11	rs1942968	C	G	.	.	.	GT	1/1
```

**Comment by freeseek** [comment]
Also, fixing the bug above exposed yet another bug in `plugins/fixref.c`. The following code:
```
else if ( args.mode==MODE_TOP2FWD )
{
    if ( ia==0 && (ib==1 || ib==2) ) // unambiguous pair: A/C or A/G
    {
        ...
    }
    else    // ambiguous pair, sequence walking must be performed
    {
        ...
    }
}
```
seems to assume that the VCF alleles are sorted in lexicographical order. I don't see why this has to be the case and it is quite unnecessary. It should be changed to something like this:
```
else if ( args.mode==MODE_TOP2FWD )
{
    int pair = 1 << ia | 1 << ib;
    if ( pair != 0x9 && pair != 0x6 )    // unambiguous pair: A/C or A/G
    {
        ...
    }
    else    // ambiguous pair, sequence walking must be performed
    {
        ...
    }
}
```

**Comment by pd3** [comment]
Thank you very much for noticing and looking into this! It is now fixed by 532cf98.
